alvr/client_openxr/src/stream.rs

```diff
@@ -366,19 +365,61 @@ impl StreamContext {
             .wait_image(xr::Duration::INFINITE)
             .unwrap();
 
+        let (flags, maybe_views) = self
+            .xr_session
+            .locate_views(
+                xr::ViewConfigurationType::PRIMARY_STEREO,
+                xr_vsync_time,
+                &self.stage_reference_space,
+            )
+            .unwrap();
+
+        let current_headset_views = if flags.contains(xr::ViewStateFlags::ORIENTATION_VALID) {
+            maybe_views
+        } else {
+            vec![crate::default_view(), crate::default_view()]
+        };
+
+        // The poses and FoVs we received from the PC runtime, which may differ and/or include
+        // altered FoVs based on settings and view conversions done for canting.
+        let input_view_params = view_params;
+        let mut output_view_params = [
+            ViewParams {
+                pose: crate::from_xr_pose(current_headset_views[0].pose),
+                fov: crate::from_xr_fov(current_headset_views[0].fov),
+            },
+            ViewParams {
+                pose: crate::from_xr_pose(current_headset_views[1].pose),
+                fov: crate::from_xr_fov(current_headset_views[1].fov),
+            },
+        ];
+        let mut openxr_display_time = vsync_time;
+
+        // (shinyquagsire23) I don't entirely trust runtimes to implement CompositionLayerProjectionView
+        // correctly, but if we do trust them, avoid doing rotation ourselves.
+        // Ex: YVR/PFDMR has issues with aspect ratio mismatches and passthrough compositing.
+        if !self.use_custom_reprojection {
+            output_view_params = input_view_params;
+
+            // Avoid passing invalid timestamp to runtime
+            // TODO(shinyquagsire23): Is there a technical reason to do it this way? Why not just vsync?
+            openxr_display_time =
+                Duration::max(timestamp, vsync_time.saturating_sub(Duration::from_secs(1)))
```

Comment:
Ideally we would like to use the correct vsync for this frame. If a frame is repeated, the vsync will happen in the past. But some runtimes will crash if the vsync we provide is too old. This seems to be a compromise that allows for isolated or small sequece of reprojected frames to keep their original vsync, but then revert to a inaccurate but valid timestamp if the freeze/reprojection continues for longer. I think this fix should be kept in the original spot, for PFD too, unless you tested that this breaks PFD support

Reply:
Or I guess phrased better, if the OpenXR runtime can't do the reprojection for us, we want as little interference as possible from the runtime if we handle it app-side. Most runtimes can just do the reprojection for us, so this is a limited edgecase.

Reply:
Right. The timestamp is actually the timestamp of the pose, which the frame should adhere to. If you use a more recent pose then you should use the timestamp of the new pose